File: app-miyou/src/main/java/com/miyou/app/domain/mission/model/UserMission.java
package com.miyou.app.domain.mission.model;

import java.time.Instant;

import com.miyou.app.domain.dialogue.model.UserId;

public record UserMission(
	UserId userId,
	MissionId missionId,
	MissionStatus status,
	Instant completedAt,
	Instant rewardedAt
) {
	public UserMission {
		if (userId == null) {
			throw new IllegalArgumentException("userId cannot be null");
		}
		if (missionId == null) {
			throw new IllegalArgumentException("missionId cannot be null");
		}
		if (status == null) {
			throw new IllegalArgumentException("status cannot be null");
		}
	}

	public static UserMission start(UserId userId, MissionId missionId) {
		return new UserMission(userId, missionId, MissionStatus.AVAILABLE, null, null);
	}

	public UserMission complete() {
		return new UserMission(userId, missionId, MissionStatus.COMPLETED, Instant.now(), null);
	}

	public UserMission reward() {
		return new UserMission(userId, missionId, MissionStatus.REWARDED, completedAt,
			Instant.now());
	}
}

File: app-miyou/src/main/kotlin/com/miyou/app/domain/mission/model/UserMission.kt
package com.miyou.app.domain.mission.model

import java.time.Instant

data class UserMission(
    val userId: String,
    val missionId: MissionId,
    val status: MissionStatus,
    val completedAt: Instant?,
    val rewardedAt: Instant?,
) {
    companion object {
        @JvmStatic
        fun start(
            userId: String,
            missionId: MissionId,
        ): UserMission = UserMission(userId, missionId, MissionStatus.AVAILABLE, null, null)
    }

    fun complete(): UserMission = copy(status = MissionStatus.COMPLETED, completedAt = Instant.now())

    fun reward(): UserMission = copy(status = MissionStatus.REWARDED, rewardedAt = Instant.now())

    fun userId(): String = userId

    fun missionId(): MissionId = missionId

    fun status(): MissionStatus = status

    fun completedAt(): Instant? = completedAt

    fun rewardedAt(): Instant? = rewardedAt
}
